### PR TITLE
DATAES-497 - Update reference documentation.

### PR DESCRIPTION
**Pull request by sothawo**
* Removed old outdated stuff
* addition of documentation and code samples for the different clients/templates/repositories
* rearranging sections to the basic outline of
  * Clients
    * Transport
    * Rest
    * Reactive
  * Operations
    * Transport
    * Rest
    * Reactive
  * Respositories
    * non-reactive
    * reactive

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAES).
- [ ] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [X] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

**Comment by sothawo** [comment]
*Note*: I kept the `TransportClient` stuff on purpose, as long as Elasticsearch does not remove it, we should keep it as well and document it.

**Comment by sothawo** [comment]
@mp911de could you please have a look at this, I think this would be good to have in the next release

**Comment by mp911de** [comment]
I will have a look later this week or Monday next week.

File: src/main/asciidoc/reference/elasticsearch-operations.adoc [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	private final ElasticsearchOperations elasticsearchOperations;
+
+	public TestController(ElasticsearchOperations elasticsearchOperations) {     <1>
+		this.elasticsearchOperations = elasticsearchOperations;
+	}
+
+	@PostMapping("/person")
+	public String save(@RequestBody Person person) {                             <2>
+
+        final IndexQuery indexQuery = new IndexQueryBuilder()
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-operations.adoc]: There's something odd with indentation. We should also remove `final` from local variable declarations.

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+.Spring Data Elasticsearch repositories using CDI
+====
+[source,java]
+----
+class ElasticsearchTemplateProducer {
+
+    @Produces
+    @ApplicationScoped
+    public ElasticsearchOperations createElasticsearchTemplate() {
+        // ... <1>
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repositories.adoc]: Let's keep callouts aligned.

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+The Spring Data Elasticsearch repositories support can be activated using an annotation through JavaConfig.
+
+.Spring Data Elasticsearch repositories using JavaConfig
+====
+[source,java]
+----
+@Configuration
+@EnableElasticsearchRepositories(                             <1>
+		basePackages = "org/springframework/data/elasticsearch/repositories"
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repositories.adoc]: Packages are delimited with `.`.

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+The Spring Data Elasticsearch module contains a custom namespace allowing definition of repository beans as well as elements for instantiating a `ElasticsearchServer` .
+
+Using the `repositories` element looks up Spring Data repositories as described in <<repositories.create-instances>> .
+
+.Setting up Elasticsearch repositories using Namespace
+====
+[source,xml]
+----
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repositories.adoc]: XML examples should make use of indentation.

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+<beans xmlns="http://www.springframework.org/schema/beans"
+xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+xmlns:elasticsearch="http://www.springframework.org/schema/data/elasticsearch"
+xsi:schemaLocation="http://www.springframework.org/schema/beans
+https://www.springframework.org/schema/beans/spring-beans-3.1.xsd
+http://www.springframework.org/schema/data/elasticsearch
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repositories.adoc]: XML examples should make use of indentation.

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc [hunk 6]
```diff
@@ -0,0 +1,157 @@
+[[elasticsearch.query-methods]]
+= Query methods
+
+[[elasticsearch.query-methods.finders]]
+== Query lookup strategies
+
+The Elasticsearch module supports all basic query building feature as String,Abstract,Criteria or have it being derived from the method name.
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: There's an issue with spacing: String,Abstract,Criteria

Not sure if it's me, I do not understand that `String,Abstract,Criteria` part.

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+=== Declared queries
+
+Deriving the query from the method name is not always sufficient and/or may result in unreadable method names. In this case one might make either use of `@Query` annotation (see <<elasticsearch.query-methods.at-query>> ).
+
+[[elasticsearch.query-methods.criterions]]
+== Query creation
+
+Generally the query creation mechanism for Elasticsearch works as described in <<repositories.query-methods>> . Here's a short example of what a Elasticsearch query method translates into:
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: Nit: Space between `>>` and `.`.

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+.Query creation from method names
+====
+[source,java]
+----
+public interface BookRepository extends Repository<Book, String>
+{
```
Comment (mp911de) [inline src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: Nit: Let's pull the curly brace one line up.

File: src/main/asciidoc/reference/elasticsearch-operations.adoc (hunk 1 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-operations.adoc]: done

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc (hunk 3 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repositories.adoc]: oops

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc (hunk 2 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repositories.adoc]: done

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc (hunk 4 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repositories.adoc]: done

File: src/main/asciidoc/reference/elasticsearch-repositories.adoc (hunk 5 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repositories.adoc]: done

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc (hunk 6 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: that was already in this form in the docs, I changed it.

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc (hunk 7 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: fixed

File: src/main/asciidoc/reference/elasticsearch-repository-queries.adoc (hunk 8 above)
Comment (sothawo) [reply src/main/asciidoc/reference/elasticsearch-repository-queries.adoc]: done

File: src/main/asciidoc/index.adoc [hunk 9]
```diff
@@ -1,9 +1,10 @@
 = Spring Data Elasticsearch - Reference Documentation
-BioMed Central Development Team; Oliver Drotbohm; Greg Turnquist; Christoph Strobl;
+BioMed Central Development Team; Oliver Drotbohm; Greg Turnquist; Christoph Strobl; Peter-Josef Meisch
 :revnumber: {version}
 :revdate: {localdate}
 :toc:
 :toc-placement!:
+//:toclevels: 5
```
[the remark this answers is not in the document]
Comment (mp911de) [reply src/main/asciidoc/index.adoc]: Checkout https://raw.githubusercontent.com/spring-projects/spring-data-mongodb/master/src/main/asciidoc/index.adoc for MongoDB and https://github.com/spring-projects/spring-data-build/blob/master/parent/pom.xml#L554-L562.

Most Asciidoc properties are set in the parent pom already. We should make the ES doc consistent with the other modules. You can build the documentation via Maven with `mvn clean package -Pdistribute`.

[the remark this answers is not in the document]
Comment (sothawo) [reply src/main/asciidoc/index.adoc]: ok, fixed it. Didn't know about the maven build, looks the same on my machine now after the build

File: src/main/asciidoc/reference/elasticsearch-object-mapping.adoc [hunk 10]
```diff
@@ -50,10 +52,10 @@ public class Config extends AbstractElasticsearchConfiguration {
 
   @Bean
   @Override
-  public EntityMapper entityMapper() { <1>
+  public EntityMapper entityMapper() {                                 <1>
```
[the remark this answers is not in the document]
Comment (mp911de) [reply src/main/asciidoc/reference/elasticsearch-object-mapping.adoc]: With `mvn clean package -Pdistribute`.

**Comment by mp911de** [comment]
Thanks a lot. That's merged and polished now.

**Pull request closed without merging**